cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -460,10 +464,9 @@ class CompositeReferenceTable {
   }
 
   template <class Builder, class PrimitiveType>
-  Result<std::shared_ptr<Array>> MaterializePrimitiveColumn(MemoryPool* memory_pool,
+  Result<std::shared_ptr<Array>> MaterializePrimitiveColumn(Builder& builder,
```

Comment:
What is this change for?

Reply:
Was fighting with the compiler for a bit on this: 
`TimestampBuilder` does not seem to be able to initialized with just a `memory_pool`, I think it requires a `arrow::DataType` object as well. That is the only difference that requires this function to be refactored and split into `MaterializePrimitveColumn` and `MaterializeTimestampPrimitiveColumn`.